### PR TITLE
firmware-management:updated: Add IPC to UpdateD

### PR DESCRIPTION
**Pull request by rwalton-arm**
UpdateD needs a way for external processes to send requests for updates.

We are using gRPC to allow us to use RPCs to communicate with UpdateD
from another process.

This patch adds gRPC as a dependency of UpdateD. We also split UpdateD
into multiple CMake projects as we now build the following executables:

* updated: our update daemon.
* updated-start-update: a gRPC client which passes the payload file path
  and the update HEADER data to use.
* updated-get-status: gRPC client which asks UpdateD for the update
  status, currently this just returns the update HEADER UpdateD was
  given when the update request was sent.

This patch also adds:

* protobuf definitions for our RPCs, and boilerplate required for
  the protoc compiler to generate C++ code from them.
* UpdateCoordinator: the core class responsible for delegating and
  tracking updates. This object has two methods called asynchronously.
  All this object does at the moment is wait and respond to our RPCs.
* Manifest: a data structure which holds the contents of the update
  header.
* PayloadHardLink: an object which manages the lifetime of the update
  payload file. UpdateD creates a hard link of the file and lets the
  filesystem do the clean-up of the underlying data when all links are
  deleted.

We also move .clang-tidy to the root of the project and disable the following
checks:
* **modernize-use-trailing-return-type** 
This is a stylistic preference for lambda style "trailing return types" which we 
don't use. 
We use regular return type annotations.

* **modernize-concat-nested-namespaces**
Another stylistic preference for namespace concatenation, we list our
namespaces as it's easier to see where the nested namespace starts
and ends.

* **fuchsia-default-arguments-calls**
Another stylistic preference, we think default arguments are fine and 
we use them.

**Jenkins** 
http://jenkins.mbed-linux.arm.com/job/rw-test/214/

**LAVA** 
http://e120856-lin.cambridge.arm.com/results.php?image_version=custom&image_number=214&custom_version=rw-test&lava_query=&lava_owner=&failures=false&submitter=robwal02&sort=by_device

**Related PRs**
https://github.com/ARMmbed/mbl-tools/pull/405 (adds UpdateD dependencies to clang-tidy checks)

### CODE REVIEW

File: firmware-management/updated/updated-client/source/rpc/Client.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/source/rpc/Client.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-client/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/CMakeLists.txt:5]: I think we should add the copyright notice and SPDX License identifier to this file.

File: firmware-management/updated/CMakeLists.txt [hunk 3]
```diff
@@ -4,19 +4,21 @@ cmake_minimum_required(VERSION 3.5)
 
 option(RUN_CODE_CHECKS OFF)
 
-if(RUN_CODE_CHECKS)
-    set(CMAKE_CXX_CLANG_TIDY clang-tidy;-checks=*,-clang-analyzer-cplusplus-*,clang-analyzer-*)
-endif(RUN_CODE_CHECKS)
-
-add_executable(updated ${CMAKE_CURRENT_SOURCE_DIR}/updated.cpp ${CMAKE_CURRENT_SOURCE_DIR}/init.cpp)
-target_include_directories(updated PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+# INTERFACE library for compile options that are not for enabling warnings.
```
Comment (diego-sueiro) [inline firmware-management/updated/CMakeLists.txt:12]: I think we should add the copyright notice and SPDX License identifier to this file.

File: firmware-management/updated/.clang-tidy [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
-Checks:          'clang-diagnostic-*,clang-analyzer-*,*,-clang-analyzer-*,clang-analyzer-cplusplus-*'
-WarningsAsErrors: 'clang-diagnostic-*,clang-analyzer-*,*,-clang-analyzer-*,clang-analyzer-cplusplus-*'
+Checks:          'clang-diagnostic-*,clang-analyzer-*,*,-clang-analyzer-alpha.*,-llvm-include-order,-modernize-use-trailing-return-type,-modernize-use-trailing-return-type,-modernize-concat-nested-namespaces,-fuchsia-default-arguments-calls'
```
Comment (diego-sueiro) [inline firmware-management/updated/.clang-tidy:2]: I think we should add the copyright notice and SPDX License identifier to this file.

File: firmware-management/updated/updated-client/source/rpc/Client.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/source/rpc/Client.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-client/source/updated-get-status.cpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/source/updated-get-status.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-client/source/updated-start-update.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/source/updated-start-update.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-rpc/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,70 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/CMakeLists.txt:5]: I think we should add the copyright notice and SPDX License identifier to this file.

File: firmware-management/updated/updated-rpc/include/updated-rpc/Error.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/include/updated-rpc/Error.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-rpc/include/updated-rpc/config.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/include/updated-rpc/config.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-rpc/source/Error.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/source/Error.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-rpc/source/config.cpp [hunk 12]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/source/config.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated-rpc/source/proto/updated-rpc.proto [hunk 13]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-rpc/source/proto/updated-rpc.proto]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/CMakeLists.txt:5]: I think we should add the copyright notice and SPDX License identifier to this file.

File: firmware-management/updated/updated/source/Manifest.h [hunk 15]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/Manifest.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/UpdateCoordinator.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/UpdateCoordinator.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/UpdateCoordinator.h [hunk 17]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/UpdateCoordinator.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/fileutils/payload.h [hunk 18]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/fileutils/payload.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/init.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/init.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/init.h [hunk 20]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/init.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/rpc/Server.cpp [hunk 21]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/rpc/Server.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/rpc/Server.h [hunk 22]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/rpc/Server.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/rpc/ServiceImpl.cpp [hunk 23]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/rpc/ServiceImpl.cpp]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

File: firmware-management/updated/updated/source/rpc/ServiceImpl.h [hunk 24]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
```
Comment (diego-sueiro) [inline firmware-management/updated/updated/source/rpc/ServiceImpl.h]:
```suggestion
 * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
```

**Review by diego-sueiro — CHANGES_REQUESTED** [review]

**Comment by diego-sueiro** [comment]
Is there any run-time testing for the IPC feature?

**Comment by rwalton-arm** [comment]
> Is there any run-time testing for the IPC feature?

You can use the `updated-start-update` and the `updated-get-status` gRPC clients to manually test the two RPCs. We don't have unit tests yet.

File: firmware-management/updated/CMakeLists.txt (hunk 3 above)
Comment (rwalton-arm) [reply firmware-management/updated/CMakeLists.txt:12]: 4c1a07c

File: firmware-management/updated/updated-client/CMakeLists.txt (hunk 2 above)
Comment (rwalton-arm) [reply firmware-management/updated/updated-client/CMakeLists.txt:5]: 4c1a07c

File: firmware-management/updated/updated-client/source/rpc/Client.h [hunk 25]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2020 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef UPDATED_CLIENT_RPC_CLIENT_H
+#define UPDATED_CLIENT_RPC_CLIENT_H
+
+#include "updated-rpc/updated-rpc.grpc.pb.h"
```
Comment (diego-sueiro) [inline firmware-management/updated/updated-client/source/rpc/Client.h:11]: Is this file auto-generated by the `protoc`? If so, I think is worth adding a comment on all includes of auto-generated headers.

**Review by diego-sueiro — CHANGES_REQUESTED** [review]

**Review by diego-sueiro — APPROVED** [review]

**Pull request merged**